tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -21,16 +21,51 @@
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
 
+namespace {
+spv::ExecutionModel getSpirvShaderKindFromHlslProfile(const char *profile) {
+  // DXIL Models are:
+  // Profile (DXIL Model) : HLSL Shader Kind : SPIRV Shader Kind
+  // vs_<version>         : Vertex Shader    : Vertex Shader
+  // hs_<version>         : Hull Shader      : Tassellation Control Shader
+  // ds_<version>         : Domain Shader    : Tessellation Evaluation Shader
+  // gs_<version>         : Geometry Shader  : Geometry Shader
+  // ps_<version>         : Pixel Shader     : Fragment Shader
+  // cs_<version>         : Compute Shader   : Compute Shader
+  switch (profile[0]) {
+  case 'v': return spv::ExecutionModel::Vertex;
+  case 'h': return spv::ExecutionModel::TessellationControl;
+  case 'd': return spv::ExecutionModel::TessellationEvaluation;
+  case 'g': return spv::ExecutionModel::Geometry;
+  case 'p': return spv::ExecutionModel::Fragment;
+  case 'c': return spv::ExecutionModel::GLCompute;
+  default:
+    assert(false && "Unknown HLSL Profile");
+    return spv::ExecutionModel::Fragment;
+  }
+}
+
+} // namespace
+
 namespace clang {
 namespace {
 
 class SPIRVEmitter : public ASTConsumer {
 public:
-  explicit SPIRVEmitter(raw_ostream *out)
-      : outStream(*out), theContext(), theBuilder(&theContext) {}
+  explicit SPIRVEmitter(CompilerInstance &ci)
+      : theCompilerInstance(ci), outStream(*ci.getOutStream()), theContext(),
+        theBuilder(&theContext) {}
 
   void HandleTranslationUnit(ASTContext &context) override {
     theBuilder.requireCapability(spv::Capability::Shader);
+    const spv::ExecutionModel shaderKind = getSpirvShaderKindFromHlslProfile(
+        theCompilerInstance.getCodeGenOpts().HLSLProfile.c_str());
+    if (shaderKind == spv::ExecutionModel::TessellationControl ||
```

Comment:
We should error out for tessellation and geometry. They are not supported yet and we don't have tests for them.